src/config/mongoose_config_parser_toml.erl

```diff
@@ -0,0 +1,354 @@
+
+process_s2s_option(<<"use_starttls">>, V) ->
+    [#local_config{key = s2s_use_starttls, value = b2a(V)}];
+process_s2s_option(<<"certfile">>, V) ->
+    [#local_config{key = s2s_cerfile, value = b2l(V)}];
+process_s2s_option(<<"default_policy">>, V) ->
+    [?HOST_F(#local_config{key = {s2s_default_policy, Host}, value = b2a(V)})];
+process_s2s_option(<<"outgoing_port">>, V) ->
+    [#local_config{key = outgoing_s2s_port, value = V}];
+process_s2s_option(<<"address">>, Addrs) ->
+    [#local_config{key = {s2s_addr, Host}, value = s2s_address(Addr)}
+     || Addr = #{<<"host">> := Host} <- Addrs].
+
+s2s_address(#{<<"ip_address">> := IP, <<"port">> := Port}) ->
+    {b2l(IP), Port};
+s2s_address(#{<<"ip_address">> := IP}) ->
+    b2l(IP).
+
+process_listener_type(Type, Content) ->
+    [process_listener(Type, C) || C <- Content].
+
+process_listener(Type, Content) ->
+    Options = maps:without([<<"port">>, <<"ip_address">>], Content),
+    PortIP = listener_portip(Content),
+    Opts = lists:flatmap(fun({K, V}) -> listener_opt(Type, K, V) end, maps:to_list(Options)),
+    {Port, IPT, _, _, Proto, OptsClean} =
+        ejabberd_listener:parse_listener_portip(PortIP, Opts),
+    {{Port, IPT, Proto}, listener_module(Type), OptsClean}.
+
+listener_portip(#{<<"port">> := Port}) -> Port;
+listener_portip(#{<<"port">> := Port, <<"ip_address">> := Addr}) -> {Port, Addr}.
+
+listener_opt(_, <<"proto">>, Proto) -> [{proto, b2a(Proto)}];
+listener_opt(_, <<"ip_version">>, 6) -> [inet6];
+listener_opt(_, <<"ip_version">>, 4) -> [inet];
+listener_opt(_, <<"backlog">>, N) -> [{backlog, N}];
+listener_opt(<<"cowboy">>, <<"tls">>, Opts) -> [{ssl, client_tls_options(Opts)}];
+listener_opt(<<"cowboy">>, <<"transport">>, Opts) ->
+    [{transport_options,
+      [{b2a(K), cowboy_transport_opt(K, V)} || {K, V} <- maps:to_list(Opts)]
+     }];
+listener_opt(<<"cowboy">>, <<"protocol">>, Opts) ->
+    [{protocol_options,
+      [{b2a(K), cowboy_protocol_opt(K, V)} || {K, V} <- maps:to_list(Opts)]
+     }];
+listener_opt(<<"cowboy">>, <<"modules">>, Modules) ->
+    [{modules,
+      [cowboy_module(Type, M) || {Type, Mods} <- maps:to_list(Modules), M <- Mods]
+     }];
+listener_opt(<<"c2s">>, <<"access">>, V) -> [{access, b2a(V)}];
+listener_opt(<<"c2s">>, <<"shaper">>, V) -> [{shaper, b2a(V)}];
+listener_opt(<<"c2s">>, <<"xml_socket">>, V) -> [{xml_socket, V}];
+listener_opt(<<"c2s">>, <<"zlib">>, V) -> [{zlib, V}];
+listener_opt(<<"c2s">>, <<"verify_peer">>, V) -> [{verify_peer, V}];
+listener_opt(<<"c2s">>, <<"hibernate_after">>, V) -> [{hibernate_after, V}];
+listener_opt(<<"c2s">>, <<"starttls">>, true) -> [starttls];
+listener_opt(<<"c2s">>, <<"starttls">>, false) -> [];
+listener_opt(<<"c2s">>, <<"starttls_required">>, true) -> [starttls_required];
+listener_opt(<<"c2s">>, <<"starttls_required">>, false) -> [];
+listener_opt(<<"c2s">>, <<"tls">>, true) -> [tls];
+listener_opt(<<"c2s">>, <<"tls">>, false) -> [];
```

Comment:
Maybe group the options that may be `false` to one clause since it returns and empty list anyway?

Reply:
I wanted to be explicit here. I think that it would be best to have one line per option. This will be refactored for sure.